titus-ext/jooq/src/main/java/com/netflix/titus/ext/jooq/jobactivity/JooqJobActivityPublisherStore.java

```diff
@@ -0,0 +1,201 @@
+
+        if (createIfNotExist) {
+            createSchemaIfNotExist();
+        }
+        queueIndex = new AtomicLong(getInitialQueueIndex());
+        logger.info("Loaded initial job activity publisher queue index {}", queueIndex);
+    }
+
+    private void createSchemaIfNotExist() {
+        dslContext.createSchemaIfNotExists(JActivity.ACTIVITY)
+                .execute();
+
+        int rc = dslContext.createTableIfNotExists(ACTIVITY_QUEUE)
+                .column(ACTIVITY_QUEUE.QUEUE_INDEX)
+                .column(ACTIVITY_QUEUE.EVENT_TYPE)
+                .column(ACTIVITY_QUEUE.SERIALIZED_EVENT)
+                .constraint(DSL.constraint("pk_activity_queue_index").primaryKey(ACTIVITY_QUEUE.QUEUE_INDEX))
+                .execute();
+        if (0 != rc) {
+            throw JobActivityException.jobActivityCreateTableException(
+                    ACTIVITY_QUEUE.getName(),
+                    new RuntimeException(String.format("Unexpected table create return code %d", rc)));
+        }
+        logger.info("Created schema and table with return code {}", rc);
+    }
+
+    private long getInitialQueueIndex() {
+        long startTime = System.currentTimeMillis();
+        Record1<Long> record = DSL.using(dslContext.configuration())
+                .select(max(ACTIVITY_QUEUE.QUEUE_INDEX))
+                .from(ACTIVITY_QUEUE)
+                .fetchOne();
+        metrics.registerSelectLatency(startTime);
+
+        // No record is present, start index at 0
+        if (null == record.value1()) {
+            return 0;
+        }
+        // Otherwise start one higher than the current max
+        return record.value1() + 1;
+    }
+
+    @VisibleForTesting
+    public Mono<Void> clearStore() {
+        return Mono.defer(() -> {
+            try {
+                int recordId = dslContext
+                        .dropTable(ACTIVITY_QUEUE)
+                        .execute();
+                logger.warn("Successfully DROP'ed table {}", ACTIVITY_QUEUE.getName(), recordId);
+            } catch (Exception e) {
+                return Mono.error(e);
+            }
+            return Mono.empty();
+        });
+    }
+
+    @VisibleForTesting
+    public long getQueueIndex() {
+        return queueIndex.get();
+    }
+
+    @Override
+    public Mono<Void> publishJob(Job<?> job) {
+        return Mono.defer(() ->
+                publishByteString(JobActivityPublisherRecord.RecordType.JOB, job.getId(), JobActivityPublisherRecordUtils.jobToByteArry(job)));
+    }
+
+    @Override
+    public Mono<Void> publishTask(Task task) {
+        return Mono.defer(() ->
+                publishByteString(JobActivityPublisherRecord.RecordType.TASK, task.getId(),
+                        JobActivityPublisherRecordUtils.taskToByteArray(task, logStorageInfo)));
+    }
+
+    private Mono<Void> publishByteString(JobActivityPublisherRecord.RecordType recordType, String recordId, byte[] serializedRecord) {
```

Comment:
The implementation is blocking. The Rx contract is also not clear. The store action should be only triggered on subscribe to the returned `Mono` object.

Reply:
Done, adjusted to move all of the APIs to transaction and wrap in a single Mono.